src/axolotl/state.rs

```diff
@@ -181,72 +180,121 @@ impl <T:Axolotl> ReceiveChain<T> {
 }
 impl <T:Axolotl> AxolotlState<T> {
 
-    pub fn encrypt(&mut self, axolotl_impl : &T, plaintext : &T::PlainText) -> (T::Message, T::Mac) {
+    pub fn encrypt(&mut self, axolotl_impl : &T, plaintext : T::PlainText) -> (T::Message, T::Mac) {
         let (new_chain_key,result) = self.encrypt_and_get_next_chain_key(axolotl_impl, plaintext);
         self.chain_key_send = new_chain_key;
         self.message_number_send += 1;
         result
     }
 
-    fn encrypt_and_get_next_chain_key(&self, axolotl_impl : &T, plaintext : &T::PlainText) -> (T::ChainKey, (T::Message, T::Mac)) {
+    fn encrypt_and_get_next_chain_key(&self, axolotl_impl : &T, plaintext : T::PlainText) -> (T::ChainKey, (T::Message, T::Mac)) {
         let (new_chain_key, message_key) = axolotl_impl.derive_next_chain_and_message_key(&self.chain_key_send);
         let ciphertext = axolotl_impl.encrypt_message(&message_key, plaintext);
 
         let message = axolotl_impl.encode_header_and_ciphertext(
-            self.message_number_send,
-            self.ratchet_key_send.public.clone(),
+            Header{
+                message_number : self.message_number_send,
+                message_number_prev : self.message_number_prev,
+                ratchet_key : self.ratchet_key_send.public.clone(),
+            },
             ciphertext
         );
-        let mac = axolotl_impl.authenticate_message(&message, &message_key, &self.identity_key_remote, &self.identity_key_local);
+        let mac = axolotl_impl.authenticate_message(&message, &message_key, &self.session_identity);
 
         (new_chain_key,(message,mac))
     }
 
-    pub fn decrypt<'a>(&mut self, axolotl_impl : &T, message : &'a T::Message, mac : T::Mac
-    ) -> Result<T::PlainText,ReceiveError<T>> where &'a T::Message : AxolotlMessageRef<T> {
-        let (message_number, message_ratchet_key) = try!(axolotl_impl.decode_header(message).map_err(|e|{ReceiveError::DecodeError(e)}));
-        let receive_chain_position =  self.receive_chains.iter().position(
-            | &ReceiveChain{ref ratchet_key, ..} | axolotl_impl.ratchet_keys_are_equal(ratchet_key, message_ratchet_key.borrow())
+    pub fn decrypt(&mut self, axolotl_impl : &T, message : T::Message, ref mac : T::Mac
+    ) -> Result<T::PlainText,ReceiveError<T>> {
+        let Header{ message_number, message_number_prev, ratchet_key : message_ratchet_key_ref } = try!(
+            axolotl_impl
+                .decode_header(&message)
+                .map_err(|e|{ReceiveError::DecodeError(e)})
         );
+        let message_ratchet_key = message_ratchet_key_ref.borrow();
+
+        self.try_decrypt_with_skipped_chain(axolotl_impl, message, mac, message_number, message_ratchet_key)
```

Comment:
This should either be simplified or pulled out into a named function. Theres much going on here to throw it into a single chained expression